modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotChecker.java

```diff
@@ -0,0 +1,698 @@
+            throw new IgniteException("No snapshot metadatas found for the baseline nodes " +
+                "with consistent ids: " + String.join(", ", baselineNodes));
+        }
+
+        return resultExceptions;
+    }
+
+    /** */
+    protected IgniteBiTuple<Map<Integer, File>, Set<File>> preparePartitions(SnapshotMetadata meta, Collection<Integer> grps, File snpDir) {
+        Map<Integer, File> grpDirs = new HashMap<>();
+        Set<File> partFiles = new HashSet<>();
+
+        Set<Integer> grpsLeft = new HashSet<>(F.isEmpty(grps) ? meta.partitions().keySet() : grps);
+
+        for (File dir : cacheDirectories(new File(snpDir, databaseRelativePath(meta.folderName())), name -> true)) {
+            int grpId = CU.cacheId(cacheGroupName(dir));
+
+            if (!grpsLeft.remove(grpId))
+                continue;
+
+            grpDirs.put(grpId, dir);
+
+            Set<Integer> parts = new HashSet<>(meta.partitions().get(grpId) == null ? Collections.emptySet()
+                : meta.partitions().get(grpId));
+
+            for (File partFile : cachePartitionFiles(dir,
+                (meta.dump() ? DUMP_FILE_EXT : FILE_SUFFIX) + (meta.compressPartitions() ? ZIP_SUFFIX : "")
+            )) {
+                int partId = partId(partFile.getName());
+
+                if (!parts.remove(partId))
+                    continue;
+
+                partFiles.add(partFile);
+            }
+
+            if (!parts.isEmpty()) {
+                throw new IgniteException("Snapshot data doesn't contain required cache group partition " +
+                    "[grpId=" + grpId + ", snpName=" + meta.snapshotName() + ", consId=" + meta.consistentId() +
+                    ", missed=" + parts + ", meta=" + meta + ']');
+            }
+        }
+
+        if (!grpsLeft.isEmpty()) {
+            throw new IgniteException("Snapshot data doesn't contain required cache groups " +
+                "[grps=" + grpsLeft + ", snpName=" + meta.snapshotName() + ", consId=" + meta.consistentId() +
+                ", meta=" + meta + ']');
+        }
+
+        return new IgniteBiTuple<>(grpDirs, partFiles);
+    }
+
+    /** */
+    public Map<PartitionKeyV2, PartitionHashRecordV2> checkSnapshotFiles(
+        File snpDir,
+        Set<Integer> grpIds,
+        SnapshotMetadata meta,
+        boolean forCreation,
+        boolean skipHash,
+        boolean punchHoleEnabled
+    ) throws IgniteCheckedException {
+        IgniteBiTuple<Map<Integer, File>, Set<File>> grpAndPartFiles = preparePartitions(meta, grpIds, snpDir);
+
+        Map<PartitionKeyV2, PartitionHashRecordV2> res = new ConcurrentHashMap<>();
+        ThreadLocal<ByteBuffer> buff = ThreadLocal.withInitial(() -> ByteBuffer.allocateDirect(meta.pageSize())
+            .order(ByteOrder.nativeOrder()));
+
+        GridKernalContext snpCtx = IgniteSnapshotManager.createStandaloneKernalContext(log, compression, snpDir, meta.folderName());
```

Comment:
There is already GridKernalContext in the class. I suppose user of the class is responsible for providing the context